service: differentiate UDP and TCP protocols

jibi opened the pull request:
This PR introduces support for protocol differentiation (UDP, TCP) for services.

The high level idea (first commit) is to take into account the L4 protocol specified in a k8s service when populating the bpf lb maps, and then use the packet's protocol to lookup services in those maps.

Next, a new `--bpf-lb-proto-diff` flag (enabled by default) is introduced. When disabled, the agent strips the protocol from the service object, so even though all the protocol differentiation logic is in place, the control plane sees only services with `NONE` protocol, effectively falling back to the old behavior. In addition to that also the datapath stops taking into account the protocol when looking up the lb maps.

Next we tackle the reason why the previous attempts didn't make it into the main branch: connection disruptions during upgrades and downgrades. The proposed solution is to keep, during upgrades, existing old style/`NONE` services as such instead of deleting them and creating new protocol-aware ones. In the control plane before creating a new service, even if protocol differentiation is enabled, we check if the same service with `NONE` protocol already exists. If that's the case, we don't create a new one. In the datapath we add an extra lookup to the lb maps: if the protocol-aware lookup fails, we set the protocol to 0 (`NONE`) and try again. This allows to go through upgrades and downgrades without disrupting connections for _existing_ `NONE` services. This doesn't solve all the cases: if a service has its protocol set and Cilium is downgraded to a version that doesn't support protocol differentiation, such connection will break as the service will be recreated.

Forth commit is just about changing `NONE` to `ANY` when dealing with no-proto services.

Finally we update all the unit and ginkgo tests to make them aware of the protocol differentiation logic, as well as the CLI to make a test aware of the protocol differentiation.

I'm currently looking into additional unit tests for controlplane and datapath, as well as additional documentation, but given that the functionality is in this PR should be already good for a first round of reviews.

jibi committed:
service: differentiate UDP and TCP protocols

this commit updates the service logic, from the agent all the way down
to the datapath, to take into account the L4 protocols specified in a
k8s service.

Historically Cilium has always been ignoring, at least from a end user
perspective, the protocol specified in a k8s service, resulting in the
inability to expose a port only for a given protocol, or in the opposite
case, resulting in a single protocol-agnostic service exposed whenever
2 TCP and UDP services on the same port were configured.

Co-authored-by: Nate Sweet <nathanjsweet@pm.me>
Co-authored-by: Antonio Ojea <aojea@google.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
service: don't overwrite existing services with NONE protocol

when upgrading to a version of Cilium that supports and has service
protocol differentiation enabled, existing services without protocol set
will be overwritten (or better, restored from the bpf maps, deleted
because they are orphan and the recreated) with their protocol set.

As this causes connection disruptions, keep those services without
protocol until the user explicitly deletes them.

To allow both type of services to co-exist, this commit:

- introduces a second service lookup in the datapath, without the
  protocol set, in case the protocol lookup fails
- updates the service allocation logic to check both for services with
  and without protocol before allocating a new one
- updates the deletion logic to look for both with and without protocol
  versions, to ensure all service versions are proper deleted

Co-authored-by: Antonio Ojea <aojea@google.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
daemon, service: add bpf-lb-proto-diff flag

add a new bpf-lb-proto-diff flag that disables the
service protocol differentiation logic, allowing to keep the old Cilium
behavior where service protocols are not distinguished

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
service: use protocol ANY when not differentiating TCP and UDP

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

aojea commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -528,6 +528,8 @@ func IsValidBackendState(state string) bool {
 
 func NewL4Type(name string) (L4Type, error) {
 	switch strings.ToLower(name) {
+	case "none":
```
self note, it is a bit late but this will be better called "any"

aojea replied:
lol, I commented before getting to this commit :)

aojea commented on pkg/service/service.go:
```diff
@@ -1344,8 +1344,25 @@ func (s *Service) createSVCInfoIfNotExist(p *lb.SVC) (*svcInfo, bool, bool,
 	prevSessionAffinity := false
 	prevLoadBalancerSourceRanges := []*cidr.CIDR{}
 
+	proto := p.Frontend.L3n4Addr.L4Addr.Protocol
```
this deserves a comment explaining the logic here and is not clear to me without having to understand the whole code base:
1. we store the original proto
2. we mutate everything to NONE
3. if does not exist with NONE with try with the original proto

did I get it right?

jibi replied:
that's correct :+1: 

> this deserves a comment explaining the logic here

yes, I'm planning to sprinkle some extra comments once I start cleaning up all this WIP commits

aojea commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -161,6 +161,7 @@ cilium-agent [flags]
       --enable-recorder                                           Enable BPF datapath pcap recorder
       --enable-route-mtu-for-cni-chaining                         Enable route MTU for pod netns when CNI chaining is used
       --enable-sctp                                               Enable SCTP support (beta)
+      --enable-service-protocol-differentiation                   TODO (default true)
```
nit, naming bikeshedding 

`--enable-service-protocol-awareness   Enable support for service protocol awareness (TCP, UDP, SCTP) ` ??

jibi replied:
I think I'll keep the current one as it's also in line with the original gh issue naming

aojea commented on daemon/cmd/daemon_main.go:
```diff
@@ -1127,6 +1127,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.StringSlice(option.NodeLabels, []string{}, "List of label prefixes used to determine identity of a node (used only when enable-node-selector-labels is enabled)")
 	option.BindEnv(vp, option.NodeLabels)
 
+	flags.Bool(option.EnableServiceProtocolDifferentiation, true, "TODO")
```
I'm not familiar with cilium config options so take this comment with that in mind.
I wonder if you want to make it the negative `--disable-service-protocol-awareness` with default `false`, users will automatically pick this value without the need of setting the flag

jibi replied:
> users will automatically pick this value without the need of setting the flag

that's already the case thanks to that second parameter set to `true`, which is the default value for the flag :+1:

aojea replied:
I see, it always pick the default you set from the flag

aojea commented on pkg/k8s/watchers/service.go:
```diff
@@ -434,6 +462,11 @@ func (k *K8sServiceWatcher) addK8sSVCs(svcID k8s.ServiceID, oldSvc, svc *k8s.Ser
 	})
 
+	if !option.Config.EnableServiceProtocolDifferentiation {
+		svc = stripServiceProtocol(svc)
```
you have `oldSvc` and `svc`, wonder if you need to modify the `oldSvc` too

jibi replied:
yep, need to look better into this :+1:

jibi replied:
> wonder if you need to modify the oldSvc too

yes, if we want to compare apples to apples `oldSvc` needs to go through `stripServiceProtocol` in tandem with `svc` :+1:

aojea commented on pkg/k8s/watchers/service.go:
```diff
@@ -209,58 +209,62 @@ func (k *K8sServiceWatcher) delK8sSVCs(svc k8s.ServiceID, svcInfo *k8s.Service)
 	})
 
-	repPorts := svcInfo.UniquePorts()
+	for _, checkProtocol := range []bool{true, false} {
```
at one point you may not need to do this as all services will be migrated so you may want to use the feature flag?

jibi replied:
I think putting this behind a feature flag would prevent deleting "old" (ANY) services after an upgrade.

Here's an example:
* user is running v1.15, so no support for service proto differentiation, and has a service deployed
* user upgrades to a Cilium version that supports service proto differentiation, and enables it (or better, doesn't explicitly disable it) 
* after the upgrade, the existing service is maintained as ANY
* at some point the user deletes the service

if this logic was put behind the feature flag, we would never be able to delete that old ANY service from Cilium state/maps, as we would only try to delete the service with the protocol set

aojea replied:
I worded it wrong, my point is that you only run this if `enable-service-protocol-differentiation` is set , that is the feature flag ... if in a long future there is no need for setting `enable-service-protocol-differentiation` because all cilium clusters are known to have been migrated, you no longer need to do this ... is just a comment

jibi replied:
ah sorry, I see what you mean now, yes, I need to add also some comments to guide us through all the code paths that need be removed once/if `enable-service-protocol-differentiation` will be deprecated :+1:

jibi committed:
service: make unit/ginkgo tests aware of service protocol differentiation

Co-authored-by: Antonio Ojea <aojea@google.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

lmb reviewed (approved):
Loader changes LGTM.

christarazi reviewed (approved):
LGTM for ci-structure

nebril reviewed (approved):
helm OK

nathanjsweet reviewed (approved):
Very clever.

derailed commented on cilium-dbg/cmd/service_update.go:
```diff
@@ -59,13 +60,38 @@ func init() {
-	frontend, err := net.ResolveTCPAddr("tcp", address)
+func parseAddress(l4Protocol, address string) (ip net.IP, port int, proto string, err error) {
+	switch proto = strings.ToLower(l4Protocol); proto {
+	case "tcp":
```
Could be out of context here but do we need to handle "sctp"?

jibi replied:
because that's part of the k8s supported protocols: https://kubernetes.io/docs/reference/networking/service-protocols/

derailed reviewed (approved):
@jibi Nice work!

viktor-kurchenko commented on .github/workflows/conformance-e2e.yaml:
```diff
@@ -417,7 +417,7 @@ jobs:
           skip-build: ${{ env.CILIUM_CLI_SKIP_BUILD }}
           image-repo: ${{ env.CILIUM_CLI_IMAGE_REPO }}
-          image-tag: ${{ env.CILIUM_CLI_VERSION }}
+          image-tag: 577aad3380b6a3dd953a7afde9cd2a8ec8383f58
```
Is this temporary change?

jibi replied:
yes, see the commit (https://github.com/cilium/cilium/pull/33434/commits/0bef665c968bec7b5324fc27c78cb3065b8eada1) title for that change

jibi committed:
cli: check: add Protocol method to LRPFrontend

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
cli: tests: lrp: add support for protocol differentiation

update the logic used to wait for local redirect bpf entries as from
v1.16 we start supporting protocol differentiation for services, which
means the frontend string description will now include also the protocol

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

dylandreimerink reviewed (approved):
Loader changes LGTM

jibi committed:
ciliumenvoyconfig: relax protocol matching for backends

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

sayboras reviewed (approved):
Thanks a lot for your help tracking down the upgrade failure in Ingress.
